gossip/src/crds_value.rs

```diff
@@ -490,84 +495,233 @@ impl Sanitize for NodeInstance {
+    len: usize,
+}
+
+impl RunLengthEncoding {
+    fn push_new_byte(
+        encoded: &mut Vec<u8>,
+        current_bit: u8,
+        current_bit_count: &mut u8,
+        len: &mut usize,
+    ) {
+        if *current_bit_count > 0 {
+            encoded.push((*current_bit_count << 1) | current_bit);
+            *len += *current_bit_count as usize;
+            *current_bit_count = 0;
+        }
+    }
+
+    pub fn new(bits: &BitVec<u8>) -> Self {
+        let mut encoded: Vec<u8> = Vec::new();
+        let mut current_bit: u8 = 1;
+        let mut current_bit_count: u8 = 0;
+        let mut len = 0;
+        for i in 0..bits.len() {
+            let bit: u8 = bits.get(i) as u8;
+            if bit == current_bit {
+                current_bit_count += 1;
+                if current_bit_count >= 0x7F {
+                    Self::push_new_byte(
+                        &mut encoded,
+                        current_bit,
+                        &mut current_bit_count,
+                        &mut len,
+                    );
+                }
+            } else {
+                Self::push_new_byte(&mut encoded, current_bit, &mut current_bit_count, &mut len);
+                current_bit = bit;
+                current_bit_count = 1;
+            }
+            if encoded.len() >= MAX_RESTART_LAST_VOTED_FORK_SLOTS_SPACE {
+                break;
+            }
+        }
+        if encoded.len() < MAX_RESTART_LAST_VOTED_FORK_SLOTS_SPACE {
+            Self::push_new_byte(&mut encoded, current_bit, &mut current_bit_count, &mut len);
+        }
+        Self { encoded, len }
+    }
+
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    pub fn to_slots(&self, last_slot: Slot, min_slot: Slot) -> Vec<Slot> {
+        let mut result: Vec<Slot> = Vec::new();
+        let mut offset = 0;
+        for byte in &self.encoded {
+            let bit_length = byte >> 1;
+            if (byte & 0x1) > 0 {
+                for i in 0..bit_length {
+                    let slot = last_slot.saturating_sub(offset).saturating_sub(i as Slot);
+                    if slot < min_slot {
+                        return result;
+                    }
+                    result.push(slot);
+                    if result.len() > MAX_SLOTS_PER_ENTRY {
+                        break;
+                    }
+                }
+            }
+            offset += bit_length as Slot;
+        }
+        result.reverse();
+        result
+    }
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq, AbiExample)]
+pub struct RawOffsets {
+    bytes: Vec<u8>,
+    len: usize,
+}
+
+impl RawOffsets {
+    pub fn new(bits: &BitVec<u8>) -> Self {
+        let mut bytes = bits.clone().into_boxed_slice().to_vec();
+        bytes.truncate(MAX_RESTART_LAST_VOTED_FORK_SLOTS_SPACE);
+        let len = min(bits.len() as usize, bytes.len() * 8);
+        Self { bytes, len }
+    }
```

Comment:
why do we need this?
why not just use the `BitVec<u8>`?

The `BitVec<u8>` can be truncated if need:
https://docs.rs/bv/latest/bv/struct.BitVec.html#method.truncate

Reply:
Tried that, serialized size is like 3x larger, I did shrink_to_fit which didn't help, any idea why?

Reply:
Never mind, changed.